### PR TITLE
Activate reverse variant of F-Split inference

### PR DESCRIPTION
**Pull request by 4tXJ7f**
This commit activates the reverse variant of the F-Split inference.

### CODE REVIEW

**Comment by 4tXJ7f** [comment]
Benchmark results (300s timeout, `--string-exp --re-elim-agg`:

```
          DIRECTORY |                       refactorProcessSimpleNEq |                             fixRevInferencesPR | 
          Benchmark |  Stat  Slvd   Tot  To Mo Err   Cpu[s]  Mem[MB] |  Stat  Slvd   Tot  To Mo Err   Cpu[s]  Mem[MB] | 
               pyex |    ok 25165 25421 256  0   0 143421.0 456616.7 |    ok 25149 25421 272  0   0 120871.8 423596.7 | 
 str_int_benchmarks |    ok 21507 21573  66  0   0  52120.3 180890.2 |    ok 21505 21573  68  0   0  54337.6 180003.9 | 
        transformed |    ee  8496  9183 249  0 438  77948.2  42051.2 |    ee  8487  9183 258  0 438  80546.5  46348.6 | 
           __totals |    ee 55168 56177 571  0 438 273489.5 679558.1 |    ee 55141 56177 598  0 438 255755.9 649949.2 | 
```

This PR performs slightly worse in terms of solved benchmarks but does better in overall solving time.

**Comment by ajreynol** [comment]
In the PR description, you should say this *adds* the reverse variant of F-Split (it didn't exist before) :)

File: src/theory/strings/core_solver.cpp [hunk 1]
```diff
@@ -1293,115 +1293,101 @@ void CoreSolver::processSimpleNEq(NormalForm& nfi,
...
-            break;
-          }
+          CVC4::String stra1 = stra.prefix(stra.size() - 1);
+          p = stra.size() - stra1.roverlap(strb);
+          Trace("strings-csp-debug") << "Compute roverlap : " << constStr << " "
+                                     << nextConstStr << std::endl;
+          size_t p2 = stra1.rfind(strb);
+          p = p2 == std::string::npos ? p : (p > p2 + 1 ? p2 + 1 : p);
+          Trace("strings-csp-debug")
+              << "overlap : " << stra1 << " " << strb << " returned " << p
```
Comment (ajreynol) [inline src/theory/strings/core_solver.cpp]: overlap -> roverlap

File: src/theory/strings/core_solver.cpp [hunk 2]
```diff
@@ -1293,115 +1293,101 @@ void CoreSolver::processSimpleNEq(NormalForm& nfi,
...
+          info.d_conc = nc.eqNode(isRev ? utils::mkNConcat(sk, prea)
+                                        : utils::mkNConcat(prea, sk));
+          info.d_new_skolem[LENGTH_SPLIT].push_back(sk);
+          info.d_id = INFER_SSPLIT_CST_PROP;
+          pinfer.push_back(info);
+          break;
+        }
       }
+
+      // Since non of the other inferences apply, we just infer that `nc` has
```
Comment (ajreynol) [inline src/theory/strings/core_solver.cpp]: non -> none

File: src/theory/strings/normal_form.cpp [hunk 3]
```diff
@@ -135,6 +136,31 @@ void NormalForm::getExplanation(int index, std::vector<Node>& curr_exp)
...
 }
 
+Node NormalForm::collectConstantStringAt(size_t& index)
+{
+  std::vector<Node> c;
+  while (d_nf.size() > index && d_nf[index].isConst())
+  {
+    c.push_back(d_nf[index]);
+    index++;
+    // break;
```
Comment (ajreynol) [inline src/theory/strings/normal_form.cpp]: What is this break comment?

**Comment by 4tXJ7f** [comment]
> In the PR description, you should say this _adds_ (or activates?) the reverse variant of F-Split (it wasn't used before) :)

Heh, good point. Done.

File: src/theory/strings/normal_form.cpp (hunk 3 above)
Comment (4tXJ7f) [reply src/theory/strings/normal_form.cpp]: Ah, I've moved the code here from the rewriter and didn't change that. Fixed.

**Comment by 4tXJ7f** [comment]
I believe, I've addressed all the comments.

**Pull request merged**